Add guide on how to create virtual fields

wysher opened the pull request:
#### Description of what you did:
Example of Virtual Fields

alexandrebodin commented:
Hey, This guide is good but the term virtual field is actually not the right one here. a virtual field is not usually persisted into DB. We could call it computed field to make it clearer imo

wysher commented:
> Hello!
> @wysher thank you for this guide, can you please add the guide into the menu by updating the config file.
> Thank you.

config updated.

wysher commented:
> Hey, This guide is good but the term virtual field is actually not the right one here. a virtual field is not usually persisted into DB. We could call it computed field to make it clearer imo

Thanks. Changed to Computed Fields

wysher commented:
> I believe we need to describe what a virtual field is as many users may not fully understand the purpose. As for the content of the guide I think it's fairly good.

@derrickmehaffy we changed name to computed fields. If in your opinion it still needs some extra purpose explanation, could You please give me some tips?

derrickmehaffy commented:
> > I believe we need to describe what a virtual field is as many users may not fully understand the purpose. As for the content of the guide I think it's fairly good.
> 
> @derrickmehaffy we changed name to computed fields. If in your opinion it still needs some extra purpose explanation, could You please give me some tips?

Yes there still should be some kind of explanation of what a computed field is, examples:

https://www.drupal.org/project/computed_field
https://hasura.io/docs/1.0/graphql/manual/schema/computed-fields.html
https://nova.laravel.com/docs/1.0/resources/fields.html#computed-fields

Something describing what it is, what it does, and any limitations of them.

wysher commented:
> > > I believe we need to describe what a virtual field is as many users may not fully understand the purpose. As for the content of the guide I think it's fairly good.
> > 
> > 
> > @derrickmehaffy we changed name to computed fields. If in your opinion it still needs some extra purpose explanation, could You please give me some tips?
> 
> Yes there still should be some kind of explanation of what a computed field is, examples:
> 
> https://www.drupal.org/project/computed_field
> https://hasura.io/docs/1.0/graphql/manual/schema/computed-fields.html
> https://nova.laravel.com/docs/1.0/resources/fields.html#computed-fields
> 
> Something describing what it is, what it does, and any limitations of them.

@derrickmehaffy Sorry, it took so long. I added this at the beginning:

```
## What are computed fields?
Computed fields are "virtual" values, dynamically added to response object. They may contain any type of data you provide based on other fields, other models fields or anything you want.
By definition they are not associated with any database column and cannot be modified directly.
```

alexandrebodin commented:
Looks good to me waiting for @derrickmehaffy  and @lauriejim reviews :)

derrickmehaffy commented on docs/v3.x/guides/computed-fields.md:
```diff
@@ -0,0 +1,102 @@
+
+## What are computed fields?
+Computed fields are "virtual" values, dynamically added to response object. They may contain any type of data you provide based on other fields, other models fields or anything you want.
+By definition they are not associated with any database column and cannot be modified directly.
```
This isn't entirely true which is why I felt a little confused, the value is being saved into a database column (as you mention this later on in the guide about disabling editing).

alexandrebodin replied:
the code was changed and it is not saved into db anymore

wysher replied:
@alexandrebodin how it changed exactly?

You said that I can remove computed field (last_name) from model, but in this case we wan't computed field to be visible in strapi admin page. Do you mean that computed field will be visible, even if it was not described in model?

alexandrebodin replied:
@wysher I didn't think you want this in the Admin panel but in the API 🤕  This is a whole other story indeed

wysher replied:
@alexandrebodin is it ok then? :)

derrickmehaffy commented on docs/v3.x/guides/computed-fields.md:
````diff
@@ -0,0 +1,102 @@
+```
+
+::: warning
+In this example we edit returned data, but data will never be saved in your database. If you want to save data, you shouldn't have Computed Fields and use beforeCreate/beforeUpdate methods instead.
````
This doesn't quite make sense to me here, as in line 50 you say to disable editing of the field.

derrickmehaffy replied:
And since you have the field defined in the model schema, there is a column created for it:

```
    "full_name": {
      "type": "string"
    }
```

alexandrebodin commented on docs/v3.x/guides/computed-fields.md:
```diff
@@ -0,0 +1,102 @@
+    "last_name": {
+      "type": "string"
+    },
+    "full_name": {
```
you can remove it

alexandrebodin commented:
Hi @wysher the guide does work but is a lot of caveats due to missing features like real virtual fields. I'm not sure this should be published as is as I wouldn't say this is standard best practices.

wysher commented:
> Hi @wysher the guide does work but is a lot of caveats due to missing features like real virtual fields. I'm not sure this should be published as is as I wouldn't say this is standard best practices.

@alexandrebodin Agreed. Should we wait until https://github.com/strapi/strapi/pull/7221 or it will not became part of strapi?

alexandrebodin commented:
@wysher We will definitely add this kind of feature. I'm just not sure about the timing :) That is something that need a bit of reflexion a an RFC I think

Tomaszal commented:
@aleksandryackovlev I've submitted [an RFC](https://github.com/strapi/rfcs/pull/17) about 2 weeks ago about this. Can some more people take a look at it since it seems to be quite inactive?

derrickmehaffy commented:
> @aleksandryackovlev I've submitted [an RFC](https://github.com/strapi/rfcs/pull/17) about 2 weeks ago about this. Can some more people take a look at it since it seems to be quite inactive?

I think you pinged the wrong alex but that's alright, he is currently out of the office until next week but I'm sure once he gets back he will review the RFCs

Tomaszal commented:
> > @aleksandryackovlev I've submitted [an RFC](https://github.com/strapi/rfcs/pull/17) about 2 weeks ago about this. Can some more people take a look at it since it seems to be quite inactive?
> 
> I think you pinged the wrong alex but that's alright, he is currently out of the office until next week but I'm sure once he gets back he will review the RFCs

Shhh, nobody saw it :sweat_smile:

alexandrebodin commented:
Hi, going to check the RFC PR when I got a minute. I think we can close this PR for now. Alright with you @wysher ?

wysher commented:
@alexandrebodin I'd rather like `hang-on` `parked` or other label, because probably we will use most of this PR in the future, but if You want we may close it. :)

alexandrebodin commented:
Closing the PR doesn't make it disappear :) I'll add a link to it in the RFC so we have a reference to it ;)

lajlev reviewed (commented):
![image](https://user-images.githubusercontent.com/133836/115523245-2af94600-a28d-11eb-9208-b3708b7dce59.png)

It is 
```
/api/client/models/client.settings.json
```
